Release topic in topic mode

zorkian opened the pull request:
Sometimes we need to release a single partition in topic claim mode for
one reason or another (failed heartbeat, somebody else claimed, etc). It
was previously possible to end up in a split brain situation where we
release partitions >0 and nobody else will ever claim them due to the
way topic claim works.

This update makes it so that if we're in topic claim mode and we release
any partition, we will kick off releases for all of them.

zorkian commented:
cc @DrTall @pingihu

DrTall commented on marshal/claim.go:
```diff
@@ -42,13 +42,14 @@ type claim struct {
 	messagesLock  *sync.Mutex
 	offsets       PartitionOffsets
 	marshal       *Marshaler
+	consumer      *Consumer
```
This up-and-down-the-stack dependency might be easier to manage going forward if you define an interface for Claim to take here instead of requiring the full-blown Consumer just to get one function? But not necessary.

DrTall commented on marshal/consumer.go:
```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
+		// This only impacts update topic claims later
+		log.Errorf("claimTerminated: failed getting current claims: %s", claimsErr)
+	}
+	delete(latestClaims, cl.topic)
```
I don't see how this is safe. We get a copy of our current claims, mutate our local copy, and only then get a lock to use that local copy to `updateTopicClaims`. If we gained a claim in the meantime before we get the lock it won't be reflected in `latestClaims`.

basharal replied:
I think the way updateTopicClaims is written today, it assumes things are being updated by a single thread since it receives the full copy. This is true because we have a single thread calling ClaimTopics today (and fast claim initially). With this implementation, since claimTerminated is called by different go routines, we can race here where calling c.GetCurrentTopicClaims followed by updateTopicClaims can race! I believe you have to take a full Lock for the duration of both operations to make sure we serialize things (in case different topics get released simultaneously).

zorkian replied:
Alright. Maybe in the future let's avoid that kind of design -- or at least strongly document it somewhere but I don't see any comments to that fact?

basharal replied:
Sounds good...I agree (apologies that it wasn't clear that it wasn't thread-safe) even though @DrTall spotted it too.

DrTall commented on marshal/consumer.go:
```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	for _, claim := range c.claims[cl.topic] {
+		if cl != claim {
```
Without this `if` we will try to double-release `cl`, is that right?

zorkian replied:
Right. Which actually is safe (it does an atomic CAS so it only ever runs once) but I figured it was easy to check (even if comparing pointers is a bit un-Go-like).

DrTall commented on marshal/consumer.go:
```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
+			if released {
+				go claim.Release()
+			} else {
+				go claim.Terminate()
```
Won't these calls to Release/Terminate result in this function getting called again immediately? We can end up trying to release each partition several times because we can release the lock before these calls to Release/Terminate take effect.

DrTall replied:
I still have this concern.

zorkian replied:
We can't ever release/terminate multiple times, since the first thing those functions do is set the closed/terminated/whatever atomic int and then the future calls are no-ops. Yeah it's possible they'll get called a couple times each but that seemed reasonable to me?

zorkian commented:
Updated cc @basharal @drtall @pingihu

I did the thing where I force pushed so you can't diff the diffs, but I did that on purpose because the new one is very much unlike the old one. I rewrote the way claimTopics gets sent. It's still a map on a channel but it's now entirely thread-safe and can be done from anywhere.

This has the side effect of slightly changing the semantics: you aren't guaranteed for one update to produced every time anything changes. It will still send updates very fast (as soon as the loop goroutine schedules) but it could possibly contain multiple updates.

I believe this behavior is fine though since the old one would drain out the updates before sending anything, leading to the same "missing" updates if you weren't consuming fast enough. That's just now happening slightly differently but it's fine.

DrTall commented on marshal/consumer.go:
```diff
@@ -622,28 +630,37 @@ func (c *Consumer) Terminate(release bool) bool {
+	}
+
+	claimedTopics := make(map[string]bool)
+	if c.Terminated() {
```
Is this necessary?

If we're properly terminated then I would think this `range` on `c.partitions` would come to the same conclusion. Namely, not adding anything to the `claimedTopics` map. However, if we are in some bogus state where we are terminated with claims perhaps it is best not to cover it up?

zorkian replied:
We can terminate w/o release, and in that case we still have claims. I am not sure it matters either way but it seemed cleaner if a terminated consumer says it owns nothing. I can revert?

DrTall commented:
Overall this looks good. I made some minor comments.

basharal commented on marshal/claim.go:
```diff
@@ -261,6 +263,13 @@ func (c *claim) teardown(releasePartition bool) bool {
+	// Advise the consumer that this claim is terminating, this is so that the consumer
+	// can release other claims if we've lost part of a topic
+	if c.consumer != nil {
+		go c.consumer.claimTerminated(c, releasePartition)
```
I know you guys like to use go routines to do stuff outside locks, but this can be bad sometimes. For example, in this case, if we call Terminate, which ends up calling this function, now, it's unclear when stuff actually get terminated because things are happening async...and will need to wait for additional things if I want to cleanup gracefully.

zorkian replied:
It is not possible to synchronously release all claims on a topic anyway. That's not really supported in the protocol. The best you could get would be to try to batch produce and assert that everything produced or not, but even then Kafka doesn't guarantee that it will only ever commit whole batches. Since the underlying protocl is asynchronous it's fine for this to be asynchronous, I believe.

How does this affect your cleanup?

basharal replied:
I think my point was mainly around when do I really know that the offsets have been committed so I can terminate the process? Doesn't this break that?

basharal commented on marshal/consumer.go:
```diff
@@ -458,76 +485,59 @@ func (c *Consumer) claimTopics() {
 
+		// Drain out any unconsumed updates
+	CONSUME:
+		for {
```
You don't need this loop if you make the topicClaimsChan of size 0 and not 1. You're guaranteed to have at most one thing in the chan

zorkian replied:
Oh! That's clever, will fix.

basharal commented on marshal/consumer_test.go:
```diff
@@ -26,22 +26,25 @@ type ConsumerSuite struct {
+		claims:             make(map[string]map[int]*claim),
+		messages:           make(chan *Message, 1000),
+		topicClaimsChan:    make(chan map[string]bool, 1),
+		topicClaimsUpdated: make(chan struct{}, 1),
```
this doesn't need to be of length 1.

basharal commented on marshal/consumer.go:
```diff
@@ -391,22 +423,21 @@ func (c *Consumer) isTopicClaimLimitReached(topic string) bool {
+
+	// Now iterate each and try to claim
+	for topic, partitions := range topic_partitions {
 		if partitions <= 0 {
```
How are you handling losing topics? This part existed to handle that case. We're comparing previous topics claimed against the current partition claims and remove things off if we no longer claim the topic.

zorkian replied:
Good catch, I need to send the update. Probably sending an update every time we hit this function (every few seconds) should be OK and that simplifies the other logic.

basharal commented on marshal/consumer.go:
```diff
@@ -458,76 +485,59 @@ func (c *Consumer) claimTopics() {
 
-// updatesTopicClaims if changed. It should be noted that it expects c.lock.RLock
-// to be already acquired
-func (c *Consumer) safeUpdateTopicClaims(latestClaims map[string]bool, force bool) {
```
it's important that we don't send the same notification multiple times. I don't believe you're breaking this behavior, but depending on how you handle losing topic claims per my comment above, this might be affected.

zorkian replied:
I don't think that's a valid constraint. It doesn't seem a reasonable design that you can't the same map twice -- the channel is just saying 'hi, here's my list of claimed topics'?

basharal replied:
I think it depends on what you're trying to communicate. The current definition, it means, topic-claims have been updated. If we're just sending periodic notification, that's no longer true. We need to double check code that is using the channel to indicate changes as opposed to periodic update and make sure they are doing the diff themselves (if they care). I will have to double check Cape's code at least.

zorkian commented:
cc @basharal @drtall update for feedback, this preserves the semantics of only sending an update when there is an update.

zorkian commented:
@basharal do you want to review this again or should Sean?

basharal commented:
let me take a second look...sorry missed it earlier

basharal commented:
LGTM...thanks a ton for doing this and for incorporating the changes to catch losing topic-claims! ship it!
